### ISSUE TITLE
sh_fault signal handler calls malloc

### ISSUE THREAD

**Issue by siteshwar** [att/ast#647]
`sh_fault` signal handler calls malloc inside `set_trap-info` function that may cause a crash:

```
0   dump_backtrace + 63
1   handle_sigsegv + 26
2   /lib64/libc.so.6(+0x346e0) [0x7fd4c39546e0]
3   __libc_malloc + 343
4   ast_malloc + 34
5   set_trapinfo + 40
6   sh_fault + 1469
7   /lib64/libc.so.6(+0x346e0) [0x7fd4c39546e0]
8   waitpid + 26
9   job_reap + 615
10  job_fork + 225
11  sh_fork + 261
12  sh_exec + 11913
13  sh_exec + 18937
14  sh_eval_20120720 + 741
15  sh_trap_20120720 + 1068
16  sh_chktrap + 1007
17  sh_exec + 30992
18  sh_exec + 22881
19  sh_exec + 19000
20  sh_exec + 18937
21  sh_exec + 24244
22  exfile + 3542
23  sh_main + 3255
24  main + 37
25  __libc_start_main + 234
26  _start + 42
```

I noticed this crash while debugging #633.

**Comment by siteshwar** [comment]
I also noticed this error which seems to be related:

```
ksh: malloc.c:2406: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_siz
e) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
/home/situ/upstream/ast/src/cmd/ksh93/tests/util/run_test.sh: line 101: 30438: Abort(coredump)
```

**Comment by krader1961** [comment]
FYI, There is this block of code in `sh_signal()` which is used to install ksh signal handlers:

```
        // Delivering a signal results in malloc() being called. If a different signal is delivered
        // while inside malloc() we can deadlock. So only allow one signal at a time with a couple
        // of exceptions. See issue #490.
        sigfillset(&sigin.sa_mask);
        sigdelset(&sigin.sa_mask, SIGABRT);
        sigdelset(&sigin.sa_mask, SIGCHLD);
        sigdelset(&sigin.sa_mask, SIGSEGV);
```

So apparently the AST team was aware of this and chose the easy way out that leaves the shell susceptible to rare failures from races involving signal delivery.

**Comment by DavidMorano** [comment]
> So apparently the AST team was aware of this and chose the easy way out that leaves the shell susceptible to rare failures from races involving signal delivery.

Yes, this issue came up in (extensive) signal testing somewhere between 10 and 15 years ago, around the time when the Solaris OS people were very active in integrating KSH into Solaris. I am not sure, but it might not even have been the Solaris people who caught these failures at the time (perhaps some other distro).  Anyway, the correct fix would have been to refactor or rewrite the KSH code to remove all calls to MALLOC from within any signal handlers, but for whatever reason (probably fear of such an extensive code change) this was never done. I was hoping that in some subsequent years, that I myself might have had the time to refactor the KSH code, but I was never able to do it myself either (due to lack of free time).  As you know, the core problem is that MALLOC is not async-signal-safe (as most things are not). MALLOC is already thread-safe and fork-safe (probably not unload-safe but that is not really needed), so MALLOC already is giving a lot to the world. In general, much old signal handling in UNIXi-like programs is very error prone (especially under high signal load), and unfortunately KSH is not an exception in this area.
